src/VisualStudio/Core/Def/Implementation/UnusedReferences/Dialog/UnusedReferencesTableProvider.DataSource.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.UnusedReferences;
+using Microsoft.VisualStudio.Shell.TableManager;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.UnusedReferences.Dialog
+{
+    internal partial class UnusedReferencesTableProvider
+    {
+        internal class UnusedReferencesDataSource : ITableDataSource
+        {
+            public const string Name = nameof(UnusedReferencesDataSource);
+
+            public string SourceTypeIdentifier => Name;
+            public string Identifier => Name;
+            public string? DisplayName => null;
+
+            private ImmutableList<SinkManager> _managers = ImmutableList<SinkManager>.Empty;
+            private ImmutableArray<UnusedReferencesEntry> _currentEntries = ImmutableArray<UnusedReferencesEntry>.Empty;
+
+            public IDisposable Subscribe(ITableDataSink sink)
+            {
+                return new SinkManager(this, sink);
+            }
+
+            public void AddTableData(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates)
+            {
+                var solutionName = Path.GetFileName(project.Solution.FilePath);
+                var entries = referenceUpdates
+                    .Select(update => new UnusedReferencesEntry(solutionName, project.Name, project.Language, update))
+                    .ToImmutableArray();
+
+                foreach (var manager in _managers)
+                {
+                    manager.Sink.AddEntries(entries);
+                }
+
+                _currentEntries = _currentEntries.AddRange(entries);
+            }
+
+            public void RemoveAllTableData()
+            {
+                foreach (var manager in _managers)
+                {
+                    manager.Sink.RemoveAllEntries();
+                }
+
+                _currentEntries = ImmutableArray<UnusedReferencesEntry>.Empty;
+            }
+
+            internal void AddSinkManager(SinkManager manager)
+            {
+                _managers = _managers.Add(manager);
+
+                manager.Sink.AddEntries(_currentEntries);
+            }
+
+            internal void RemoveSinkManager(SinkManager manager)
+            {
+                _managers = _managers.Remove(manager);
+
+                manager.Sink.RemoveAllEntries();
```

Comment:
Is this necessary? I wasn't sure if this would surprise the disposer of the subscription since they might not be expecting any further calls.

Reply:
It isn't necessary just being paranoid. The behavior is the same with or without it.